runtime/near-vm-runner/src/wasmer2_runner.rs

```diff
@@ -301,62 +271,245 @@ pub(crate) fn wasmer2_vm_hash() -> u64 {
+                if f.signature.params().is_empty() && f.signature.results().is_empty() {
+                    let trampoline = f.call_trampoline.expect("externs always have a trampoline");
+                    // SAFETY: we double-checked the signature, and all of the remaining arguments
+                    // come from an exported function definition which must be valid since it comes
+                    // from wasmer itself.
+                    wasmer_vm::wasmer_call_trampoline(
+                        self,
+                        f.vmctx,
+                        trampoline,
+                        f.address,
+                        [].as_mut_ptr() as *mut _,
+                    )
+                    .map_err(|e| {
+                        translate_runtime_error(RuntimeError::from_trap(e), import.vmlogic)
+                    })?;
+                } else {
+                    panic!("signature should've already been checked by `get_entrypoint_index`")
+                }
+            } else {
+                panic!("signature should've already been checked by `get_entrypoint_index`")
+            }
 
-    let import = imports::wasmer2::build(store, memory_copy, &mut logic, current_protocol_version);
+            {
+                let _span =
+                    tracing::debug_span!(target: "vm", "run_method/drop_instance").entered();
+                drop(instance)
+            }
+        }
 
-    if let Err(e) = check_method(module, method_name) {
-        return (None, Some(e));
+        Ok(())
     }
 
-    let err = run_method(module, &import, method_name, &mut logic).err();
-    (Some(logic.outcome()), err)
+    pub(crate) fn run_module<'a>(
+        &self,
+        artifact: &VMArtifact,
+        memory: &mut Wasmer2Memory,
+        method_name: &str,
+        ext: &mut dyn External,
+        context: VMContext,
+        fees_config: &'a RuntimeFeesConfig,
+        promise_results: &'a [PromiseResult],
+        current_protocol_version: ProtocolVersion,
+    ) -> (Option<VMOutcome>, Option<VMError>) {
+        let vmmemory = memory.vm();
+        let mut logic = VMLogic::new_with_protocol_version(
+            ext,
+            context,
+            &self.config,
+            fees_config,
+            promise_results,
+            memory,
+            current_protocol_version,
+        );
+        let import = imports::wasmer2::build(vmmemory, &mut logic, current_protocol_version);
+        if let Err(e) = get_entrypoint_index(&*artifact.artifact, method_name) {
+            return (None, Some(e));
+        }
+        let err = self.run_method(artifact, import, method_name).err();
+        (Some(logic.outcome()), err)
+    }
 }
 
-pub(crate) struct Wasmer2VM;
+unsafe impl wasmer_vm::TrapHandler for Wasmer2VM {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn custom_trap_handler(&self, _: &dyn Fn(&wasmer_vm::TrapHandlerFn) -> bool) -> bool {
+        false
+    }
+}
```

Comment:
Feels like something we potentially don't need :)

Reply:
Right, this trait is required due to trait bounds on `wasmer_call_trampoline`. I could split out `Wasmer2VM` into two types, but this seemed less onerous to me.